src/compiler/checker.ts

```diff
@@ -7272,22 +7300,61 @@ module ts {
                 links.type = instantiateType(getTypeOfSymbol(lastOrUndefined(context.parameters)), mapper);
             }
         }
+        
+        function createPromiseType(promisedType: Type): Type {
+            // creates a `Promise<T>` type where `T` is the promisedType argument
+            let globalPromiseType = getGlobalPromiseType();
+            if (globalPromiseType !== emptyObjectType) {
+                // if the promised type is itself a promise, get the underlying type; otherwise, fallback to the promised type
+                promisedType = getAwaitedType(promisedType);
+                if (promisedType !== unknownType) {
+                    return createTypeReference(<GenericType>globalPromiseType, [promisedType]);
+                }
+            }
+            
+            return undefined;
+        }
 
         function getReturnTypeFromBody(func: FunctionLikeDeclaration, contextualMapper?: TypeMapper): Type {
             let contextualSignature = getContextualSignatureForFunctionLikeDeclaration(func);
             if (!func.body) {
                 return unknownType;
             }
+            
+            let isAsync = isAsyncFunctionLike(func);
             let type: Type;
             if (func.body.kind !== SyntaxKind.Block) {
-                type = checkExpressionCached(<Expression>func.body, contextualMapper);
+                if (isAsync) {
+                    // From within an async function you can return either a non-promise value or a promise. Any 
+                    // Promise/A+ compatible implementation will always assimilate any foreign promise, so the 
+                    // return type of the body is not the return type of the consise body, rather it

```

Comment:
semicolon rather than comma; I am also unsure what you mean by "concise body" unless you just mean "body".